File: app/src/main/java/xyz/b515/schedule/db/CourseManager.java
package xyz.b515.schedule.db;

import android.content.Context;
import android.util.Log;

import java.util.List;

import xyz.b515.schedule.entity.Course;
import xyz.b515.schedule.entity.Spacetime;

public class CourseManager {
    private static final String TAG = "CourseManager";

    private DBHelper dbHelper;

    public CourseManager(Context context) {
        this.dbHelper = new DBHelper(context);
    }

    public boolean insertCourse(List<Course> courses) {
        try {
            dbHelper.getCourseDao().create(courses);
            return true;
        } catch (Exception e) {
            Log.e(TAG, "insertCourse", e);
        }
        return false;
    }

    public boolean insertCourse(Course course) {
        try {
            dbHelper.getCourseDao().create(course);
            return true;
        } catch (Exception e) {
            Log.e(TAG, "insertCourse", e);
        }
        return false;
    }

    public List<Course> getAllCourse() {
        try {
            return dbHelper.getCourseDao().queryForAll();
        } catch (Exception e) {
            Log.e(TAG, "getAllCourse", e);
        }
        return null;
    }

    public Course getCourse(String name) {
        try {
            return dbHelper.getCourseDao().queryBuilder().where().eq("name", name).queryForFirst();
        } catch (Exception e) {
            Log.e(TAG, "getCourse", e);
        }
        return null;
    }

    public int updateCourse(Course course) {
        try {
            return dbHelper.getCourseDao().update(course);
        } catch (Exception e) {
            Log.e(TAG, "updateCourse", e);
        }
        return 0;
    }

    public boolean clearCourse() {
        try {
            dbHelper.getCourseDao().deleteBuilder().delete();
            dbHelper.getSpacetimeDao().deleteBuilder().delete();
            return true;
        } catch (Exception e) {
            Log.e(TAG, "clearCourse", e);
        }
        return false;
    }

    public boolean insertSpacetime(Spacetime spacetime) {
        try {
            dbHelper.getSpacetimeDao().create(spacetime);
            return true;
        } catch (Exception e) {
            Log.e(TAG, "insertSpacetime", e);
        }
        return false;
    }

    public int updateSpacetime(Spacetime spacetime) {
        try {
            return dbHelper.getSpacetimeDao().update(spacetime);
        } catch (Exception e) {
            Log.e(TAG, "updateSpacetime", e);
        }
        return 0;
    }

    public boolean deleteCourse(Course course){
        try {
            dbHelper.getCourseDao().delete(course);
            dbHelper.getSpacetimeDao().delete(course.getSpacetimes());
            return true;
        } catch (Exception e) {
            Log.e(TAG, "deleteCourse", e);
        }
        return false;
    }

    public boolean deleteSpacetime(Spacetime spacetime){
        try {
            dbHelper.getSpacetimeDao().delete(spacetime);
            return true;
        } catch (Exception e) {
            Log.e(TAG, "deleteSpacetime", e);
        }
        return false;
    }
}

File: app/src/main/kotlin/xyz/b515/schedule/db/CourseManager.kt
package xyz.b515.schedule.db

import android.content.Context
import android.util.Log
import com.j256.ormlite.dao.Dao

import xyz.b515.schedule.entity.Course
import xyz.b515.schedule.entity.Spacetime

class CourseManager(context: Context) {
    private val TAG = "CourseManager"
    private val dbHelper = DBHelper(context)

    private fun courseTransaction(body: (Dao<Course, Int>) -> Int): Int {
        try {
            return body(dbHelper.courseDao)
        } catch (e: Exception) {
            Log.e(TAG, body.toString(), e)
        }
        return 0
    }

    private fun <T> courseTransaction(default: T, body: (Dao<Course, Int>) -> T?): T {
        try {
            return body(dbHelper.courseDao) ?: default
        } catch (e: Exception) {
            Log.e(TAG, body.toString(), e)
        }
        return default
    }

    private fun <T> courseTransaction(body: (Dao<Course, Int>) -> MutableList<T>): MutableList<T> {
        try {
            return body(dbHelper.courseDao)
        } catch (e: Exception) {
            Log.e(TAG, body.toString(), e)
        }
        return mutableListOf()
    }

    private fun spacetimeTransaction(body: (Dao<Spacetime, Int>) -> Int): Int {
        try {
            return body(dbHelper.spacetimeDao)
        } catch (e: Exception) {
            Log.e(TAG, body.toString(), e)
        }
        return 0
    }

    private fun courseSpacetimeTransaction(body: (Dao<Course, Int>, Dao<Spacetime, Int>) -> Int): Int {
        try {
            return body(dbHelper.courseDao, dbHelper.spacetimeDao)
        } catch (e: Exception) {
            Log.e(TAG, body.toString(), e)
        }
        return 0
    }

    fun getAllCourse() = courseTransaction<Course> { it.queryForAll() }

    fun insertCourse(courses: List<Course>) = courseTransaction { it.create(courses) }

    fun insertCourse(course: Course) = courseTransaction { it.create(course) }

    fun getCourse(name: String) = courseTransaction<Course>(Course()) {
        it.queryBuilder().where().eq("name", name).queryForFirst()
    }

    fun getCourse(id: Int) = courseTransaction<Course>(Course()) { it.queryForId(id) }

    fun updateCourse(course: Course) = courseTransaction { it.update(course) }

    fun clearCourse() = courseSpacetimeTransaction { cd, sd ->
        cd.deleteBuilder().delete() + sd.deleteBuilder().delete()
    }

    fun insertSpacetime(spacetime: Spacetime) = spacetimeTransaction { it.create(spacetime) }

    fun updateSpacetime(spacetime: Spacetime) = spacetimeTransaction { it.update(spacetime) }

    fun deleteCourse(course: Course) = courseSpacetimeTransaction { cd, sd ->
        cd.delete(course) + sd.delete(course.spacetimes)
    }

    fun deleteSpacetime(spacetime: Spacetime) = spacetimeTransaction { it.delete(spacetime) }

}
